### ISSUE TITLE
Can we get rid of the unencrypted /boot partition in Q4.1?

### ISSUE THREAD

**Issue by qwrd** [QubesOS/qubes-issues#6151]
Hello,

A while ago I decided to give Qubes 4.1 a test run, and I noticed there were now three partitions created on the disk by the setup. First one was an EFI partition, second one was an unencrypted /boot partition and the last was LUKS-encrypted rootfs.

So basically I've gotten into this habit of modifying all my linux systems to use just two disk partitions - an EFI partition for the bootloader entry and an encrypted rootfs partition (with /boot being part of it). This is possible because grub2 supports decrypting the LUKS rootfs and loading an initramfs image from the /boot folder on the encrypted partition. The procedure was (AFAIK) first described by Pavel Kogan in the following entries:

https://www.pavelkogan.com/2014/05/23/luks-full-disk-encryption/
https://www.pavelkogan.com/2015/01/25/linux-mint-encryption/

**So basically what the grub2 bootloader does is:**
- Asks for passphrase 
- Decrypts the rootfs
- Loads the initramfs into memory from the encrypted /boot folder
- Uses a keyfile from initramfs to unlock the LUKS rootfs (so the user isn't prompted for password twice)
- Boots the system

I've been exploring the possibility of modifying my disk layout in the same manner for Qubes 4.1 because I noticed that Q4.1 now uses grub2 as its bootloader. So I was wondering, if it is feasible (or rather in any interest to the Qubes community) to have the setup modified in this manner. The drawback would be that we lose the nice graphical password entry UI and the fact that this grub2 early stage only supports the US-based keyboard layout.

As of now I believe grub2 only supports decryption of LUKS1 format in this manner, but I have noticed that support for LUKS2/argon2 is being worked on in the master.

https://savannah.gnu.org/bugs/?55093
https://lists.gnu.org/archive/html/grub-devel/2020-03/msg00109.html

Thoughts?

**Comment by iamahuman** [comment]
If you boot through EFI only, the /boot paritition is not required AFAIK.

**Comment by qwrd** [comment]
Hello.

In the classical LUKS encryption scheme for a linux distribution the /boot partition was **required** to remain unencrypted because it hosted the grub2 configuration files, initramfs image as well as the kernel. The recent EFI grub2 image (at least the one in debian or ubuntu - see package grub-efi-amd64-signed) has an ability to decrypt a LUKS rootfs and read those files from the /boot folder inside the encrypted partition, hence eliminating the need to store those files in an unencrypted disk space.

That is unless you're talking about storing those files on the EFI partition itself, but that doesn't gain any additional security since the files will still be unencrypted. The Qubes 4.1 setup does create a separate /boot partition by default, if choosing to encrypt the filesystem during setup.

**Comment by marmarek** [comment]
> If you boot through EFI only, the /boot paritition is not required AFAIK.

It is (on Qubes 4.1), to load kernel and initramfs, no longer _copied_ to EFI partition, as we have grub2-efi back.

As for using encrypted /boot and keyfile embedded info initramfs, there are a couple of issues:
 - grub2 does not support LUKS2 (which we use)
 - we want to integrate https://github.com/osresearch/safeboot/issues/21 at some point - which also require unencrypted initramfs (but not necessary separate unencrypted /boot, just EFI partition)

**Comment by qwrd** [comment]
**marmarek:**
- Recent grub2 does support LUKS2, but it didn't fully support the new Argon2 PBKDF yet, so you had to use the old pbkdf2 instead. But Argon2 support was being worked on last time I checked the mailing list.

- Interesting, I'll keep an eye out on the development. Though I kind of am a fan of having as much parts of the boot process encrypted as possible...

~q

**Comment by marmarek** [comment]
safeboot allows you to have boot binaries signed, which is much more important than having them encrypted. Besides luks keyfile introduced into initramfs to avoid double password prompt, there are really no secrets in /boot - those are just files from packages available publicly. What you really want is to assure integrity of /boot, not necessary confidentiality (encryption is some approximation of that, but [not a perfect one](https://stackoverflow.com/questions/3655516/does-encryption-guarantee-integrity)). Furthermore, if you encrypt just /boot, then one can still mess with the stage1 of grub2 which still needs to be unencrypted (BIOS/UEFI needs to load it somehow) - attacker can simply modify that part to capture your disk passphrase, inject rootkit or whatever else, instead of doing that with files in /boot.
Ideally, the system should authenticate to you (prove to you it wasn't modified) _before_ you enter any secret, including passphrase prompt. This is how [AEM](https://www.qubes-os.org/doc/anti-evil-maid/) works ([background info](https://blog.invisiblethings.org/2011/09/07/anti-evil-maid.html), which I really recommend to read before considering boot security solutions). Moving that prompt earlier (into grub) makes such verification much harder (of possible at all), as you have much less tools for that in grub.

In short: what you want is integrity protection of boot files, not necessary confidentiality. Encrypting /boot gives you just small part of it (better than none, but far from what is needed and makes it harder to implement the proper thing).

**Comment by andrewdavidwong** [comment]
Duplicate of #2442

**Comment by andrewdavidwong** [comment]
This appears to be a [duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) of an existing issue. If so, please comment on the appropriate existing issue instead. If you believe this is not really a duplicate, please leave a comment briefly explaining why. We'll be happy to take another look and, if appropriate, reopen this issue. Thank you.
